tests/artifacts_tests/test_download_artifact.py

```diff
@@ -0,0 +1,41 @@
+from __future__ import annotations
+
+import pathlib
+
+import pytest
+
+import optuna
+from optuna.artifacts import download_artifact
+from optuna.artifacts import FileSystemArtifactStore
+from optuna.artifacts import upload_artifact
+from optuna.artifacts._protocol import ArtifactStore
+
+
+@pytest.fixture(params=["FileSystem"])
+def artifact_store(tmp_path: pathlib.PurePath, request: pytest.FixtureRequest) -> ArtifactStore:
+    if request.param == "FileSystem":
+        return FileSystemArtifactStore(str(tmp_path))
+    assert False, f"Unknown artifact store: {request.param}"
+
+
+def test_download_artifact(tmp_path: pathlib.PurePath, artifact_store: ArtifactStore) -> None:
+    storage = optuna.storages.InMemoryStorage()
+    study = optuna.create_study(storage=storage)
+    artifact_ids: list[str] = []
+
+    def objective(trial: optuna.Trial) -> float:
+        x = trial.suggest_int("x", 0, 100)
+        y = trial.suggest_int("y", 0, 100)
+        dummy_file = str(tmp_path / f"dummy_{trial.number}.txt")
+        with open(dummy_file, "w") as f:
+            f.write(f"{x} {y}")
+        artifact_ids.append(upload_artifact(trial, dummy_file, artifact_store))
+        return x**2 + y**2
+
+    study.optimize(objective, n_trials=5)
+
+    for i, artifact_id in enumerate(artifact_ids):
+        dummy_downloaded_file = str(tmp_path / f"dummy_downloaded_{i}.txt")
+        download_artifact(artifact_store, artifact_id, dummy_downloaded_file)
+        with open(dummy_downloaded_file, "r") as f:
+            assert f.read() == f"{study.trials[i].params['x']} {study.trials[i].params['y']}"
```

Comment:
I'm not sure, but can we use temporary files/directories to keep the environment clean?

Reply:
Dummy files actually remain on `tmp_path`, so the machine where this test is run would not be kept completely clean.
Since I implemented this test referring to `test_upload_artifact`, which does not keep the environment clean as well, we may need to modify both tests if we do not want this situation.

Reply:
Sorry for the delayed response.
Temporary files can sometimes complicate the execution of git commands. So, we might not want to leave them to minimize operational mistakes.

> Since I implemented this test referring to `test_upload_artifact`, which does not keep the environment clean as well, we may need to modify both tests if we do not want this situation.

Thank you for pointing it out. For the scope of this PR, let us solely concentrate on the tests of downloading, and we can work on the other files in follow-up PRs.

Reply:
According to [this](https://github.com/pytest-dev/pytest/issues/8141#issuecomment-1278960826), pytest keeps 3 temporary directories by default, and we can prevent pytest from saving temporary directories by setting `tmp_path_retention_policy`.
So, adding just one line `tmp_path_retention_policy = "none"` to `pyproject.toml` would resolve this problem. Should I include this fix in this PR?

Reply:
My understanding is that there are two concerns, neither of which is a problem, so I think this PR is fine as is.
The first is that temporary files complicate git management, but since `tmp_path` is created in a system temporary directory, it does not affect git. (https://docs.pytest.org/en/7.3.x/how-to/tmp_path.html#the-default-base-temporary-directory)
The second is that the temporary files affect other test runs, but since `tmp_path` is unique for each function invocation, this is not a problem. (https://docs.pytest.org/en/7.3.x/reference/reference.html#tmp-path)

Reply:
Ah, I misunderstood. These test cases employ the `tmp_path` fixture instead of creating temporary files in the test cases. Thanks for your explanation.